[c10d] Make alltoall as a custom op

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79779
* #79777
* __->__ #79691

Summary:
This patch makes alltoall as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_cuda
    BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_cuda_complex
    BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_full_group_cuda
    and other existing distributed tests.

Commit:
[c10d] Make alltoall as a custom op

Summary:
This patch makes alltoall as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
Existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make alltoall as a custom op"


Summary:
This patch makes alltoall as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
Existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make alltoall as a custom op"


Summary:
This patch makes alltoall as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_cuda
    BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_cuda_complex
    BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_full_group_cuda
    and other existing distributed tests.

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -97,6 +97,19 @@ c10::intrusive_ptr<ProcessGroup::Work> scatter_(
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> alltoall_(
+    at::TensorList output_tensors,
```
why the type here is different from other comm ops?

mrshenli replied:
not stamping yet due to this comment

alanwaketan replied:
I think it's a convention to use TensorList to represent const std::vector<at::Tensor>&.

I only use const std::vector<at::Tensor>& if there is a const std::vector<std::vector<at::Tensor>>& in the signature to keep consistency on that function. Let me know which way you like.

mrshenli replied:
I see. can we add some docs to explain that in the code? Thanks!

alanwaketan commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1263,18 +1263,25 @@ that adds a prefix to each key inserted to the store.
           .def(
               "alltoall",
-              [](::c10d::ProcessGroup& pg,
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& pg,
```
Let me follow up on https://github.com/pytorch/pytorch/issues/80246.

alanwaketan commented on torch/csrc/distributed/c10d/Ops.hpp:
```diff
@@ -29,6 +29,10 @@ TORCH_API c10::intrusive_ptr<ProcessGroup::Work> scatter(const c10::intrusive_pt
       const std::vector<std::vector<at::Tensor>>& input_tensors,
       const ScatterOptions& opts ={});
+TORCH_API c10::intrusive_ptr<ProcessGroup::Work> alltoall(const c10::intrusive_ptr<ProcessGroup>& process_group,
+      at::TensorList output_tensors,
```
The pull request passes the linter. And the format seems consistent with other function signatures in the same file. What's your concern here?

mrshenli replied:
We used to run clang-format on all distributed cpp files. Not sure what's today's convention. Below is what clang-format gives me:

```c++
namespace c10d {
namespace ops {

// Below are essentially ProcessGroup's corresponding ops but routed to the
// dispatcher.
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
broadcast(const c10::intrusive_ptr<ProcessGroup> &process_group,
          at::TensorList tensors, const BroadcastOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
allreduce(const c10::intrusive_ptr<ProcessGroup> &process_group,
          at::TensorList tensors, const AllreduceOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
allgather(const c10::intrusive_ptr<ProcessGroup> &process_group,
          const std::vector<std::vector<at::Tensor>> &output_tensors,
          const std::vector<at::Tensor> &input_tensors,
          const AllgatherOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
reduce_scatter(const c10::intrusive_ptr<ProcessGroup> &process_group,
               const std::vector<at::Tensor> &output_tensors,
               const std::vector<std::vector<at::Tensor>> &input_tensors,
               const ReduceScatterOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
reduce(const c10::intrusive_ptr<ProcessGroup> &process_group,
       at::TensorList tensors, const ReduceOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
gather(const c10::intrusive_ptr<ProcessGroup> &process_group,
       const std::vector<std::vector<at::Tensor>> &output_tensors,
       const std::vector<at::Tensor> &input_tensors,
       const GatherOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
scatter(const c10::intrusive_ptr<ProcessGroup> &process_group,
        const std::vector<at::Tensor> &output_tensors,
        const std::vector<std::vector<at::Tensor>> &input_tensors,
        const ScatterOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
alltoall(const c10::intrusive_ptr<ProcessGroup> &process_group,
         at::TensorList output_tensors, at::TensorList input_tensors,
         const AllToAllOptions &opts = {});

} // namespace ops
} // namespace c10d
```

alanwaketan replied:
Somehow, my local linter doesn't produce the same output as yours. Let me just copy yours.

mrshenli reviewed (approved):
LGTM. Left a minor comment

Commit:
Update on "[c10d] Make alltoall as a custom op"


Summary:
This patch makes alltoall as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_cuda
    BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_cuda_complex
    BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_full_group_cuda
    and other existing distributed tests.

[ghstack-poisoned]

alanwaketan commented:
@pytorchbot merge --green

alanwaketan commented:
Thanks for approving this pull request, Shen and Wanchao.
